### Level 1:
* Finish the first level by finding the best attribute.
    * We calculated the Information Gain for Meat.
    * There are two other attributes to check
* Split on the best attribute – This is your level one attribute test.
### Level 2:
* Find the best attribute for the left most node at the second level and
split the node accordingly
* Assume sub-nodes are sorted alphabetically left to right by attribute name
* Label any leaf nodes with their majority class
* You could/should continue with the other nodes to get more practice

In [17]:
from typing import List
import numpy as np
import pandas as pd

In [48]:
tennis_data = {
    # "Day": ["D1", "D2", "D3", "D4", "D5", "D6", "D7", "D8", "D9", "D10", "D11", "D12", "D13", "D14"],
    "Outlook": ["Sunny", "Sunny", "Overcast", "Rain", "Rain", "Rain", "Overcast", "Sunny", "Sunny", "Rain", "Sunny", "Overcast", "Overcast", "Rain"],
    "Temp": ["Hot", "Hot", "Hot", "Mild", "Cool", "Cool", "Cool", "Mild", "Cool", "Mild", "Mild", "Mild", "Hot", "Mild"],
    "Humidity": ["High", "High", "High", "High", "Normal", "Normal", "Normal", "High", "Normal", "Normal", "Normal", "High", "Normal", "High"],
    "Wind": ["Weak", "Strong", "Weak", "Weak", "Weak", "Strong", "Strong", "Weak", "Weak", "Weak", "Strong", "Strong", "Weak", "Strong"],
    "Play Tennis": ["No", "No", "Yes", "Yes", "Yes", "No", "Yes", "No", "Yes", "Yes", "Yes", "Yes", "Yes", "No"]
}

pizza_data = [
    {'ID': 'e1', 'Meat': 'yes', 'Crust': 'Thin',    'Veg': 'no',  'Quality': 'Great'},
    {'ID': 'e2', 'Meat': 'no',  'Crust': 'Deep',    'Veg': 'no',  'Quality': 'Bad'},
    {'ID': 'e3', 'Meat': 'no',  'Crust': 'Stuffed', 'Veg': 'yes', 'Quality': 'Good'},
    {'ID': 'e4', 'Meat': 'yes', 'Crust': 'Stuffed', 'Veg': 'yes', 'Quality': 'Great'},
    {'ID': 'e5', 'Meat': 'yes', 'Crust': 'Deep',    'Veg': 'no',  'Quality': 'Good'},
    {'ID': 'e6', 'Meat': 'yes', 'Crust': 'Deep',    'Veg': 'yes', 'Quality': 'Great'},
    {'ID': 'e7', 'Meat': 'no',  'Crust': 'Thin',    'Veg': 'yes', 'Quality': 'Good'},
    {'ID': 'e8', 'Meat': 'yes', 'Crust': 'Deep',    'Veg': 'no',  'Quality': 'Good'},
    {'ID': 'e9', 'Meat': 'no',  'Crust': 'Thin',    'Veg': 'no',  'Quality': 'Bad'},
]


df = pd.DataFrame(pizza_data)
display(df)

# # display(df.value_counts("Play Tennis"))
classifier_index = -1
feature_index = 2
# feature_col = df.columns[feature_index]
# df[feature_col].count()


df_classifier = df.iloc[:, classifier_index]
df_class_counts = df_classifier.value_counts()

display(df_class_counts)
print(df_class_counts.tolist())

ID Meat    Crust  Veg Quality
0  e1  yes     Thin   no   Great
1  e2   no     Deep   no     Bad
2  e3   no  Stuffed  yes    Good
3  e4  yes  Stuffed  yes   Great
4  e5  yes     Deep   no    Good
5  e6  yes     Deep  yes   Great
6  e7   no     Thin  yes    Good
7  e8  yes     Deep   no    Good
8  e9   no     Thin   no     Bad

Quality
Good     4
Great    3
Bad      2
Name: count, dtype: int64

[4, 3, 2]


In [47]:
# def entropy(pos: int, neg: int):
#     if pos == neg:
#         return 1
    
#     if pos == 0 or neg == 0:
#         return 0

#     total = pos + neg

#     return -(pos / total) * np.log2(pos / total) - (neg / total) * np.log2(neg / total)

def entropy_calc(counts: List[int]):
    sum: int = 0
    entropy: List[float] = []
    total: float = 0

    for count in counts:
        sum += count

    for count in counts:
        if (count == 0):
            continue
        total -= count / sum * np.log2(count / sum)

    for element in entropy:
        total += element
    return total

def gain(df: pd.DataFrame, classifier_index: int = -1, feature_index: int = 0):
    
    df_classifier = df.iloc[:, classifier_index]
    df_class_counts = df_classifier.value_counts()

    # display(df_class_counts)
    
    overall_entropy = entropy_calc(df_class_counts.tolist())
    # sunny_days_count = df[df['outlook'] == 'sunny'].shape[0]
    # sunny_play_yes_count = df[(df['outlook'] == 'sunny') & (df['play'] == 'yes')].shape[0]
    # sunny_play_no_count = df[(df['outlook'] == 'sunny') & (df['play'] == 'no')].shape[0]

    # eval_list = []
    gain_total = overall_entropy

    feature_col = df.columns[feature_index]
    classifier_col = df.columns[classifier_index]
    feature_values = df[feature_col].unique()
    class_labels = df[classifier_col].unique()
    dataset_total = df[feature_col].count()

    print("---------------------------------")

    for val in feature_values:
        feature_counts: List[int] = []
        sub_df = df[df[feature_col] == val]
        class_counts = sub_df[classifier_col].value_counts()
        internal_sum = 0
        # print(f"Class counts {class_counts}")
        print(f"Feature Value: '{val}")
        for label in class_labels:
            count = class_counts.get(label, 0)
            print(f"Label {label}: {count}")
            internal_sum += count
            feature_counts.append(count)
        
        gain_total -= internal_sum / dataset_total * entropy_calc(feature_counts)


    return gain_total



    # overall_entropy - 

    # return overall_entropy

print(f"Meat Output: {gain(df, classifier_index=-1, feature_index=1)}")
print(f"Crust Output: {gain(df, classifier_index=-1, feature_index=2)}")
print(f"Veg Output: {gain(df, classifier_index=-1, feature_index=3)}")

# entropy(2, 3)

---------------------------------
Feature Value: 'yes
Label Great: 3
Label Bad: 0
Label Good: 2
Feature Value: 'no
Label Great: 0
Label Bad: 2
Label Good: 2
Meat Output: 0.5466316153937776
---------------------------------
Feature Value: 'Thin
Label Great: 1
Label Bad: 1
Label Good: 1
Feature Value: 'Deep
Label Great: 1
Label Bad: 1
Label Good: 2
Feature Value: 'Stuffed
Label Great: 1
Label Bad: 0
Label Good: 1
Crust Output: 0.11328333429487492
---------------------------------
Feature Value: 'no
Label Great: 1
Label Bad: 2
Label Good: 2
Feature Value: 'yes
Label Great: 2
Label Bad: 0
Label Good: 2
Veg Output: 0.24053300404228117


Meat wins at a gain score of 0.5466316153937776

Current Tree:

- Meat
    - No
    - Yes

In [55]:
sub_df = df[df['Meat'] == 'no']

display(sub_df)

print(f"No Meat + Crust: {gain(sub_df, classifier_index=-1, feature_index=2)}")
print(f"No Meat + Veg: {gain(sub_df, classifier_index=-1, feature_index=3)}")
# print(f"Has Meat + Wind: {gain(sub_df, classifier_index=-1, feature_index=3)}")


ID Meat    Crust  Veg Quality
1  e2   no     Deep   no     Bad
2  e3   no  Stuffed  yes    Good
6  e7   no     Thin  yes    Good
8  e9   no     Thin   no     Bad

---------------------------------
Feature Value: 'Deep
Label Bad: 1
Label Good: 0
Feature Value: 'Stuffed
Label Bad: 0
Label Good: 1
Feature Value: 'Thin
Label Bad: 1
Label Good: 1
No Meat + Crust: 0.5
---------------------------------
Feature Value: 'no
Label Bad: 2
Label Good: 0
Feature Value: 'yes
Label Bad: 0
Label Good: 2
No Meat + Veg: 1.0


Vegitables wins at a gain score of 1

Current Tree:

- Meat
    - No: Veg
        - No: Pizza is Bad
        - Yes: Pizza is Good
    - Yes

In [62]:
sub_df = df[df['Meat'] == 'yes']

display(sub_df)

print(f"Has Meat + Crust: {gain(sub_df, classifier_index=-1, feature_index=2)}")
print(f"Has Meat + Veg: {gain(sub_df, classifier_index=-1, feature_index=3)}")
# print(f"Has Meat + Wind: {gain(sub_df, classifier_index=-1, feature_index=3)}")


ID Meat    Crust  Veg Quality
0  e1  yes     Thin   no   Great
3  e4  yes  Stuffed  yes   Great
4  e5  yes     Deep   no    Good
5  e6  yes     Deep  yes   Great
7  e8  yes     Deep   no    Good

---------------------------------
Feature Value: 'Thin
Label Great: 1
Label Good: 0
Feature Value: 'Stuffed
Label Great: 1
Label Good: 0
Feature Value: 'Deep
Label Great: 1
Label Good: 2
Has Meat + Crust: 0.4199730940219749
---------------------------------
Feature Value: 'no
Label Great: 1
Label Good: 2
Feature Value: 'yes
Label Great: 2
Label Good: 0
Has Meat + Veg: 0.4199730940219749


Tie Between Crust and Veg, so Crust will win because `C` comes before `V` in the alphebet

Current Tree:

- Meat
    - No: Veg
        - No: Pizza is Bad
        - Yes: Pizza is Good
    - Yes: Crust
        - Thin: Pizza is Great
        - Stuffed: Pizza is Great
        - Deep: 

In [64]:

# A subset of the Meat: yes and Crust: deep
display(sub_df)

meat_yes_deep_crust_df = sub_df[sub_df['Crust'] == 'Deep']

display(sub_df)

print(f"Has Meat + Deep Crust + Veg: {gain(meat_yes_deep_crust_df, classifier_index=-1, feature_index=3)}")

ID Meat Crust  Veg Quality
4  e5  yes  Deep   no    Good
5  e6  yes  Deep  yes   Great
7  e8  yes  Deep   no    Good

ID Meat Crust  Veg Quality
4  e5  yes  Deep   no    Good
5  e6  yes  Deep  yes   Great
7  e8  yes  Deep   no    Good

---------------------------------
Feature Value: 'no
Label Good: 2
Label Great: 0
Feature Value: 'yes
Label Good: 0
Label Great: 1
Has Meat + Deep Crust + Veg: 0.9182958340544896


For any pizza that both has meat and deep crust, if it has veggies,
then it will be great, otherwise it will just be good

Final Tree:

- Meat
    - No: Veg
        - No: Pizza is Bad
        - Yes: Pizza is Good
    - Yes: Crust
        - Deep: Veg
            - No: Pizza is Good
            - Yes: Pizza is Great
        - Stuffed: Pizza is Great
        - Thin: Pizza is Great